File: node/features/step_definitions/sync_inject.steps.js
const { spawnSync } = require('child_process');
const fs = require('fs');
const os = require('os');
const path = require('path');
const assert = require('assert');
const { Given, When, Then, setWorldConstructor } = require('@cucumber/cucumber');

const SKILL_REL_PATH = path.join('.agent-skills', 'project-management-with-beads', 'SKILL.md');
const packagedPath = path.join(__dirname, '..', '..', 'data', 'skills', 'project-management-with-beads', 'SKILL.md');

class World {
  constructor() {
    this.tmpRoot = fs.mkdtempSync(path.join(os.tmpdir(), 'beads-skill-'));
    this.repo = path.join(this.tmpRoot, 'repo');
    fs.mkdirSync(this.repo);
    fs.mkdirSync(path.join(this.repo, '.git'));
  }
}
setWorldConstructor(World);

function run(cmd, args, cwd) {
  const r = spawnSync(cmd, args, { stdio: 'inherit', cwd });
  if (r.status !== 0) throw new Error(`${cmd} ${args.join(' ')} failed`);
}

Given('a git-initialized temporary repository', function () {
  // repo already created in world
});

When('I run beads-skill sync', function () {
  run('node', [path.join(__dirname, '..', '..', 'bin/cli.js'), 'sync', '--repo', this.repo], this.repo);
});

When('I run beads-skill inject', function () {
  run('node', [path.join(__dirname, '..', '..', 'bin/cli.js'), 'inject', '--repo', this.repo], this.repo);
});

Then('the skill file matches the packaged bytes', function () {
  const target = path.join(this.repo, SKILL_REL_PATH);
  assert.ok(fs.existsSync(target), 'skill file missing after sync');
  const packagedHash = fs.readFileSync(packagedPath);
  assert.strictEqual(
    fs.readFileSync(target, 'utf8'),
    packagedHash.toString('utf8'),
    'synced skill differs from packaged content'
  );
});

Then('injecting again leaves AGENTS.md unchanged', function () {
  const agents = path.join(this.repo, 'AGENTS.md');
  const before = fs.readFileSync(agents, 'utf8');
  run('node', [path.join(__dirname, '..', '..', 'bin/cli.js'), 'inject', '--repo', this.repo], this.repo);
  assert.strictEqual(fs.readFileSync(agents, 'utf8'), before, 'inject not idempotent');
});

File: python/features/steps/sync_inject_steps.py
import hashlib
import subprocess
import sys
from pathlib import Path

import pytest
from pytest_bdd import given, scenario, then, when

import agent_skills.cli as cli


@pytest.mark.bdd
@scenario("../sync_inject.feature", "Sync writes skill and inject is idempotent", features_base_dir=None)
def test_sync_inject():
    """BDD scenario runner"""


@given("a git-initialized temporary repository", target_fixture="repo")
def repo(tmp_path: Path) -> Path:
    repo_path = tmp_path / "repo"
    repo_path.mkdir()
    (repo_path / ".git").mkdir()
    return repo_path


@when("I run agent-skills sync")
def run_sync(repo: Path):
    subprocess.check_call([sys.executable, "-m", "agent_skills.cli", "sync", "--repo", str(repo)])


@when("I run agent-skills inject")
def run_inject(repo: Path):
    subprocess.check_call([sys.executable, "-m", "agent_skills.cli", "inject", "--repo", str(repo)])


@then("the skill file matches the packaged bytes")
def skill_matches(repo: Path):
    skill_path = repo / cli.SKILL_REL_PATH / cli.SKILL_FILE_NAME
    assert skill_path.exists(), "skill file missing after sync"
    packaged_hash = hashlib.sha256(cli._skill_bytes()).hexdigest()
    assert hashlib.sha256(skill_path.read_bytes()).hexdigest() == packaged_hash


@then("injecting again leaves AGENTS.md unchanged")
def inject_idempotent(repo: Path):
    agents = repo / "AGENTS.md"
    before = agents.read_text()
    subprocess.check_call([sys.executable, "-m", "agent_skills.cli", "inject", "--repo", str(repo)])
    assert agents.read_text() == before
